src/utils/ERC20CosmosCoinAmountConversion.sol

```diff
@@ -0,0 +1,136 @@
+    /**
+     * @notice Gets the decimals of a token if it extends the IERC20 standard using IERC20Metadata
+     * @param tokenAddress The address of the token contract
+     * @return decimals The number of decimals of the token
+     */
+    function _getERC20TokenDecimals(address tokenAddress) internal view returns (uint8) {
+        // Input validation
+        // TODO Add custom error?
+        require(tokenAddress != address(0), "Address cannot be the zero address");
+        // TODO Reason about this check - Remember it as a good practice but what real protections give us?
+        // TODO Add custom error?
+        require(tokenAddress != address(this), "Address cannot be the contract itself");
+
+        // If the tokens extends the IERC20 it should be using IERC20Metadata which supports the decimals() call
+        // Why this? -->  https://detectors.auditbase.com/decimals-erc20-standard-solidity
+        try IERC20Metadata(tokenAddress).decimals() returns (uint8 decimals) {
+            return decimals;
+        } catch {
+            // If the decimals function is not implemented, return the default value
+            return DEFAULT_ERC20_DECIMALS;
+        }
+    }
+    // Convert ERC20 tokens decimals to Cosmos decimals units
+    // NOTE that the refund to the user of the remainder of the conversion should happen during the transfer.
+    /**
+     * @notice Convert the uint256 ERC20 token amount into cosmos coin amount uint64
+     * @notice This functions allows us to consider the loss of precision
+     * @param tokenAddress The address of the token contract
+     * @param amount The amount to be converted
+     * @return convertedAmount The amount converted to uint64 supported by cosmos coins
+     * @return remainder The remainder of the conversion
+     */
+
+    function _convertERC20AmountToCosmosCoin(
+        address tokenAddress,
+        uint256 amount
+    )
+        internal
+        view
+        returns (uint64 convertedAmount, uint256 remainder)
+    {
+        // Note if we want to allow different cosmos decimals, we need to handle that here too
+        // Note that tokenAddress input validations are executed in the _getERC20TokenDecimals function
+        uint8 tokenDecimals = _getERC20TokenDecimals(tokenAddress);
+        // TODO write an ADR for this?
+        require(tokenDecimals >= 6, "ERC20 Tokens with less than 6 decimals are not supported");
+        // TODO Add custom error?
+        // Amount input validation
+        require(amount != 0, "Requested conversion for the 0 amount");
+        // Ensure the amount respects the token's decimals
+        // Handle the case where the input amount exceeds the token's precision
+        uint256 temp_convertedAmount;
+        uint256 factor;
+        // Case ERC20 decimals are bigger than cosmos decimals
+        if (tokenDecimals > DEFAULT_COSMOS_DECIMALS) {
+            factor = 10 ** (tokenDecimals - DEFAULT_COSMOS_DECIMALS);
```

Comment:
Are there any scenarios where this can cause any overflows?

Reply:
Solidity version starting from [0.8](https://docs.soliditylang.org/en/latest/080-breaking-changes.html) supports built-in underflows and overflows checks for arithmetic operations. We should be good!